Verification Case API

yunjoonjung-PNNL opened the pull request:
Verification Case API implementation.

I'll continue making other ```VerificationClass``` methods progress on top of this work. Thank you!

yunjoonjung-PNNL committed:
finish verification case constructor

yunjoonjung-PNNL committed:
add missing comment in the constructor

yunjoonjung-PNNL committed:
add type hint in read_case method

yunjoonjung-PNNL committed:
finish load_verification_cases_from_json method development

yunjoonjung-PNNL committed:
move verification_case_unit_test.json to 'data' dir

yunjoonjung-PNNL committed:
fix load_verification_cases_from_json unit tests

yunjoonjung-PNNL committed:
fix assert equal in the unit test

yunjoonjung-PNNL committed:
fix JSON file path in unit test

yunjoonjung-PNNL committed:
finish save_case_suite_to_json development

yunjoonjung-PNNL committed:
fix the path in test_save_case_suite_to_json_check_file_saving

yunjoonjung-PNNL committed:
include the unit test outcome files

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,167 @@
+        """
+
+        self.case_suite = {}
+        if case is not None:
```
what if case is None and file_path is not None (or the other way around). The handling logic has issues here, especially if the design is to have both optional.

yunjoonjung-PNNL replied:
Even if one of the arguments is ```None```, isn't ```not None``` argument read correctly?

leijerry888 replied:
Sorry, I did not say it correctly. What if both are not None (and both are valid), or both are not None and only one of the two is valid).

yunjoonjung-PNNL replied:
I don't think I fully understood your point. 

From my understanding,
1. When both are ```not None``` and valid, the data in the arguments are recorded in ```self.case_suite``` dict.

3. When both are ```not None``` and only one of the two is valid, the code causes an error for the invalid argument.

leijerry888 replied:
as we chatted, we just need to make sure it is clear in the design and documentation to the user what the intended behavior is: we exit whenever any argument is illegal, and when they are both legal, we merge cases from both source together.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,167 @@
+
+        return case_suite
+
+    def load_verification_cases_from_json(self, json_case_path: str) -> list:
```
List?

leijerry888 replied:
Please don't directly resolve it without code change or response. Similar to below

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,167 @@
+        """Instantiate a verification case class object and load verification case(s) in `self.case_suite` as a Dict. keys are automatically generated unique id of cases, values are the fully defined verification case Dict.
+
+        Args:
+            case: case dictionary that includes verification items.
```
Please write docstring with types properly formatted. Otherwise, documentation won't be correctly generated later. This applies to all occasions.

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,167 @@
+                f"The type of the 'json_case_path' argument has to be str, but {type(json_case_path)} type is provided. Please verify the 'json_case_path' argument."
+            )
+
+        return list(self.case_suite.keys())
```
We should only return keys of newly added cases, otherwise, these cases are kinda lost. With error, we should either return None or an empty list

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,167 @@
+        if len(case_ids) == 0:
+            if len(self.case_suite) != 0:
+                for case in self.case_suite.items():
+                    case_suite_in_template_format_append(case[1])
```
I suggest we also save the case id in the case json element as a reference. key name can be "case_id_in_suite"

yunjoonjung-PNNL replied:
I agree with the point, but I'm a bit confused about the implementation.

To verify, is it correct to add ```case_id_in_suite``` key in the verification case dict along with some other keys such as ```no```, ```run_simulation```, ```simulation_IO```?

leijerry888 replied:
yes, put it at root level of the case is fine.

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,167 @@
+            return None
+
+        # if json_path is saved in a different dir, check if it exists
+        json_path_split = json_path.split("/")
+        dir_path = "/".join(json_path_split[:-1])
+        if len(dir_path) > 2:  # when json_path includes directory path
+            if not os.path.exists(dir_path):
+                logging.error(
+                    f"{dir_path} directory doesn't exist. Please make sure to provide an already existing directory."
+                )
+                return None
```
several potential issues here:
1. we did not check if this is a json file path
2. why not checking dir_path validity when its len <= 2
3. we assume / s are used in the path but this is not always the truth

if we just want to report error when there if no file. you can catch the exception directly or just let python to handle it.
if we want to create the folder if it does not exist, then additional things need to be added

yunjoonjung-PNNL replied:
1. Added the JSON file path-checking part.
2. When ```len(dir_path) <= 2```, it means the ```json_case_path``` doesn't include dir (only includes file name e.g., testing.json).
Example: 
i) if json_path = "testing.json",  dir_path = ".", len(dir_path ) = 1
ii) if json_path = "./testing.json",  dir_path = "", len(dir_path ) = 0
iii) if json_path3 = "./testing1/testing.json", dir_path = "./testing1", len(dir_path ) = **10**

This part is further updated as shown below.
```python
dir_path = json_path.split("/")[-1:]
```
This makes the above part to
i) if json_path = "testing.json",  dir_path = ".", len(dir_path ) = 1
ii) if json_path = "./testing.json",  dir_path = "", len(dir_path ) = 0
iii) if json_path3 = "./testing1/testing.json",  dir_path = "./testing1", len(dir_path ) = **2**


3. I believe the case when ```/```s aren't used is also considered.

leijerry888 replied:
to point 2. what about "." or "./" or "/"?

yunjoonjung-PNNL replied:
Great point.

Based on your discussion last time, I realized that this approach would cause some issues if users use different os (windows, mac, linux) because as far as know each os deals with paths a bit differently (e.g., /, //, \\, etc.). Hence, I'll leave out the logic about checking whether a directory exists or not. and let Python deal with this.

Please let me know if you have a different thought.

yunjoonjung-PNNL replied:
Great point.

Based on your discussion last time, I realized that this approach would cause some issues if users use different os (windows, mac, Linux) because as far as know each os deals with paths a bit differently (e.g., /, //, \, etc.). Hence, I'll leave out the logic about checking whether a directory exists or not. and let Python deal with this.

Please let me know if you have a different thought.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,167 @@
+        # check case_ids type
+        if not isinstance(case_ids, list):
+            logging.error(
+                f"The type of the 'case_ids' argument has to be str, but {type(case_ids)} type is provided. Please verify the 'case_ids' argument."
```
list, not str

yunjoonjung-PNNL replied:
Addressed.

yunjoonjung-PNNL committed:
finish validate_verification_case_structure method development

yunjoonjung-PNNL committed:
mark [x] to the validate_verification_case_structure method

yunjoonjung-PNNL committed:
black formatting

yunjoonjung-PNNL committed:
finish save_verification_cases_to_json development

yunjoonjung-PNNL committed:
finish create_verificaton_case_suite_from_base_case development

yunjoonjung-PNNL committed:
combine test_save_case_suite_to_json unit test methods

yunjoonjung-PNNL committed:
combine test_load_verification_cases_from_json method unit tests

yunjoonjung-PNNL committed:
combine validate_verification_case_structure method unit tests

yunjoonjung-PNNL committed:
add to use validate_verification_case_structure method in other methods

yunjoonjung-PNNL committed:
add missing comment

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,458 @@
+                if case_id in self.case_suite:
+                    case_suite_in_template_format_append(self.case_suite[case_id])
+                else:
+                    logging.error(
```
Is this an error or warning? if it is error, we need return

yunjoonjung-PNNL replied:
This is an error, so I added ```return None``` here.

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,458 @@
+        Args:
+            json_path: str. json file path to save the cases.
+            cases: List. List of complete verification cases Dictionary to save.
+            keep_base_case: bool. whether to keep the base case in returned list of verification cases. Default to False.
```
is it default to true of false? in LOC 172 it is True. here it is False

yunjoonjung-PNNL replied:
The default should be False. Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,458 @@
+
+        Args:
+            json_path: str. json file path to save the cases.
+            cases: List. List of complete verification cases Dictionary to save.
```
this parameters do not match the def

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,458 @@
+    @staticmethod
+    def create_verificaton_case_suite_from_base_case(
+        base_case: Dict, update_key_value: Dict, keep_base_case: bool = True
+    ):
```
add return typing

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,528 @@
+                "idf": ["Testing_file1.idf", "Testing_file2.idf"],
+                "idd": "../Energy+V9_0_1.idd",
+                "weather": "./USA_GA_Atlanta-Hartsfield.Jackson.Intl.AP.722190_TMY3.epw",
+                "output": "eplusout.csv",
```
we don't want unchanged keys and values to be included here

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,458 @@
+
+                elif isinstance(value, list):
+                    for idx, case in enumerate(casees):
+                        if key in [
+                            "no",
+                            "run_simulation",
+                            "expected_result",
+                            "datapoints_source",
+                            "verification_class",
+                        ]:
+                            case[key] = value[idx]
+                        elif key in ["idf", "idd", "weather", "output", "ep_path"]:
+                            case["simulation_IO"][key] = value[idx]
+
+                        elif key == "subject":
+                            case["datapoints_source"]["idf_output_variables"][
+                                prev_level_key
+                            ]["subject"] = value[idx]
+
+                        elif key == "variable":
+                            case["datapoints_source"]["idf_output_variables"][
+                                prev_level_key
+                            ]["variable"] = value[idx]
+
+                        elif key == "frequency":
+                            case["datapoints_source"]["idf_output_variables"][
+                                prev_level_key
+                            ]["frequency"] = value[idx]
+
+                        elif prev_level_key == "parameters":
+                            case["datapoints_source"]["parameters"][key] = value[idx]
```
this part looks very hard-coded. I feel there is no need to do this, just follow the key value structure of the base case will be fine. I did not look carefully here because other part of the code and tests need to be updated first. If it is not clear what to do, we can revisit this part once other updates are done

yunjoonjung-PNNL replied:
This part is further optimized. If there is a better way of doing this, I believe we can revisit this part once other parts are updated.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,458 @@
+                            if not isinstance(
+                                instance[key][idf_key].get("subject"), str
+                            ):
+                                if verbose:
+                                    logging.error(
+                                        f"The type of '{idf_key}' subject's value must be str, but {type(instance[key][idf_key].get('subject'))} is provided."
+                                    )
+                                return False
+
+                            if not instance[key][idf_key].get("variable"):
+                                if verbose:
+                                    logging.error(
+                                        f"'variable' key doesn't exist in {instance[key][idf_key]}. Please make sure to include the 'variable' key."
+                                    )
+                                return False
+
+                            if not isinstance(
+                                instance[key][idf_key].get("variable"), str
+                            ):
+                                if verbose:
+                                    logging.error(
+                                        f"The type of '{idf_key}' variable's value must be str, but {type(instance[key][idf_key].get('variable'))} is provided."
+                                    )
+                                return False
+
+                            if not instance[key][idf_key].get("frequency"):
+                                if verbose:
+                                    logging.error(
+                                        f"'frequency' key doesn't exist in {instance[key][idf_key]}. Please make sure to include the 'frequency' key."
+                                    )
+                                return False
+
+                            if not isinstance(
+                                instance[key][idf_key].get("frequency"), str
+                            ):
+                                if verbose:
+                                    logging.error(
+                                        f"The type of '{idf_key}' frequency's value must be str, but {type(instance[key][idf_key].get('frequency'))} is provided."
+                                    )
+                                return False
```
to avoid hard coding, I suggest checking structure integrity against an external definition (something similar to the json schema template)

yunjoonjung-PNNL replied:
This part is further optimized. 

Question) How do we check structure integrity against an external definition? Could you be a bit more specific?

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,528 @@
+from api import VerificationCase
+
+
+class TestDataProcessing(unittest.TestCase):
```
This comment applies to all unit tests below: correct returns of the methods also need to be tested

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,458 @@
+                else:
+                    if key in instance:
+                        if type(instance[key]) != schema[key]:
+                            if verbose:
```
verbose is not to control error handling. Error is error no matter verbose or not. verbose is to give detailed info (not err) to user.

leijerry888 replied:
this applies to all similar cases below

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,458 @@
+            "expected_result": str,
+            "datapoints_source": {"idf_output_variables": dict, "parameters": dict},
+            "verification_class": str,
+        }
```
this looks odd. no need to define a hard coded dict here for checking. Please think together with the method below. Check against a schema requirement defined in file will be better.

yunjoonjung-PNNL replied:
The dict is used in the ```_validate_case_structure_helper``` method to help examine each value's type.

Do we have the schema requirement? Could you direct me to where it is?

leijerry888 replied:
I understand that. Two points for you to consider. 1) If it is only used in _validate_case_structure_helper, compared with defining it here and serving it to _validate_case_structure_helper as a parameter, maybe moving it inside _validate_case_structure_helper (thus no need to pass in a parameter) will be better ; 2) we likely would want to change schema structure requirements, defining it in a separate json file will make updating it easier than modifying python code.

yunjoonjung-PNNL replied:
I agree with both (1) and (2). Let me put this JSON in the ```schema``` directory.

yunjoonjung-PNNL replied:
Updates

Originally, I was going to create a JSON file for this, but I decided not to.
Reason)
1. in JSON file, values should be str (```"int"``` instead of ```int```). For this reason, there has to be one more for-loop for type conversion.
2. In the future, I believe this part will be covered with JSON case schema that we will develop in the near future so, we will eliminate this part anyway later.

leijerry888 replied:
we actually only have string, bool and numbers (int or float) in the case definition, so the mapping would be minimal. The current form is a bit more hard-coded than I would like and we can think about this later.

yunjoonjung-PNNL replied:
sounds great!

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,458 @@
+        return _validate_case_structure_helper(case_schema, case, verbose)
+
+    @staticmethod
+    def save_verification_cases_to_json(json_path: str, cases: list):
```
save_case_suite_to_json should be using this instead of implementing it again.

yunjoonjung-PNNL replied:
The thing is the ```save_case_suite_to_json``` method takes list of ```case_ids```, and the ```save_verification_cases_to_json``` method takes ```cases```, as shown below. In this case, can't we use the ```save_case_suite_to_json``` method in the ```save_verification_cases_to_json``` method?

```python
def save_case_suite_to_json(self, json_path: str, case_ids: list = []):
```


```python
@staticmethod
def save_verification_cases_to_json(json_path: str, cases: list):
```

leijerry888 replied:
from coding perspective, you can not call instance method in static method.

from logic perspective, save case suite to json is essentially saving verification cases to json where cases being a subset of self.case_suites

yunjoonjung-PNNL replied:
Yes, but I believe the issue is the ```save_verification_cases_to_json``` method has the option to save cases selectively by having the ```case_ids``` arg (if the list is empty, save all cases), and the ```save_verification_cases_to_json``` method doesn't. Hence, if we use the ```save_verification_cases_to_json``` method in the ```save_case_suite_to_json``` method, how can we save a few specific cases based on ```case_ids``` arg?

leijerry888 replied:
I don't think I understand the question. My point is to have save_case_suite_to_json pick cases based on case_ids and call save_verification_cases_to_json to save those picked cases.

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,528 @@
+        )
+        file_path = "./tests/api/result/from_test_save_case_suite_to_json_correct_json_path.json"
+        vc.save_case_suite_to_json(file_path)
+        assert os.path.isfile(file_path)
```
delete this file after asssert. and remove them from repository. Also apply to similar cases

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,528 @@
+        assert os.path.isfile(file_path)
+
+        # test the given case is saved correctly.
+        case_dict = {
```
this is not used below

yunjoonjung-PNNL replied:
Fixed (Deleted).

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,528 @@
+        }
+        saving_file_path = "./tests/api/result/from_test_save_case_suite_to_json_check_file_saving.json"
+        vc = VerificationCase(
+            case=None, file_path="./tests/api/data/verification_case_unit_test.json"
```
we should allow object instantiation without any cases.

yunjoonjung-PNNL replied:
I believe this is possible. Please refer to the first comment.

leijerry888 reviewed (changes requested):
Thanks for the PR!

In addition to specific comments to code, a few general comments for you to consider:

High priority todo: unit tests need to include tests of correct functionalities. I did not check the coding logic carefully, I will wait till these tests implemented to do so as that will make it much easier to do. Thanks.

Then,
1. We talked about assign None to all arguments so we can handle exception more flexible. Please refer to other PRs for this
2. Please double check your doc string format and completeness (including default values, correct capitalization of types, align with methods definitions, etc.)

yunjoonjung-PNNL committed:
address PR comments

yunjoonjung-PNNL committed:
fix return typehint

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+
+
+class VerificationCase:
+    def __init__(self, case: Dict = None, file_path: str = None) -> NoReturn:
```
__init__ should return None, not NoReturn. NoReturn is when it never return anything.

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+                    case["case_id_in_suite"] = str(unique_hash)
+                    self.case_suite[uuid.uuid1()] = case
+            else:
+                return None
```
this means when case is not valid, return None without error. Not what we want and does not align with the docstring "If one argument is invalid, the class is terminated with an error message regardless of the other arguments validity."

yunjoonjung-PNNL replied:
The error msg comes from the ```check_type``` method. The ```return None``` was used to stop the code.

leijerry888 replied:
Got it. Putting return None here is not thorough if you want to use it to break the code. E.g. if evaluating LOC 20  is False, there is no return None in this block. I suggest you combine these conditions into one check, maybe that will make it easier.

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+            logging.error(
+                f"The `{var_name}` argument's type must be {var_type}, but {type(var_value)} is provided."
+            )
+            return None
```
maybe return False is better than return None, thus check_type will just return a bool, which logically makes more sense to me

yunjoonjung-PNNL replied:
Great point! Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+                    # create a case_suite consisting of key: unique id, value: verification case
+                    unique_hash = uuid.uuid1()
+                    case["case_id_in_suite"] = str(unique_hash)
+                    self.case_suite[uuid.uuid1()] = case
```
uuid.uuid1() will give a new uuid, should use unique_hash

yunjoonjung-PNNL replied:
good catch! Fixed.

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,546 @@
+        case_original = copy.deepcopy(case)
+        vc = VerificationCase(case=case, file_path=None)
+        vs_case_suite = copy.deepcopy(vc.case_suite.popitem()[1])
+        del vs_case_suite[
```
Don't delete the uuid. Check uuid in case is the actual uuid in its key after fixing the code

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,546 @@
+
+        # test when the wrong `case` arg is provided
+        with self.assertLogs() as logobs:
+            testing_case = [
+                {
+                    "cases": {
+                        "no": 1,
+                        "run_simulation": True,
+                        "simulation_IO": {
+                            "idf": "../test_cases/doe_prototype_cases/ASHRAE901_Hospital_STD2019_Atlanta_Case1.idf",
+                            "idd": "../resources/Energy+V9_0_1.idd",
+                            "weather": "../weather/USA_GA_Atlanta-Hartsfield.Jackson.Intl.AP.722190_TMY3.epw",
+                            "output": "eplusout.csv",
+                            "ep_path": "C:\\EnergyPlusV9-0-1\\energyplus.exe",
+                        },
+                        "expected_result": "pass",
+                        "datapoints_source": {
+                            "idf_output_variables": {
+                                "T_sa_set": {
+                                    "subject": "VAV_1 Supply Equipment Outlet Node",
+                                    "variable": "System Node Setpoint Temperature",
+                                    "frequency": "detailed",
+                                }
+                            },
+                            "parameters": {},
+                        },
+                        "verification_class": "Testing",
+                    }
+                }
+            ]  # wrong data type
+            vc = VerificationCase(case=testing_case, file_path=None)
```
Only allow loading one case as a dict in constructor is weird. Loading a list of dict feels more reasonable. As we discussed earlier, maybe not allow any case loading in constructor is a better / cleaner.

yunjoonjung-PNNL replied:
I agree with changing to load case**s** (a list of dict).

Should we make one new method that does what the current constructor does?

leijerry888 replied:
no. We should always expect a list of cases, even if there is only one case. change the parameter name to cases

yunjoonjung-PNNL replied:
Ok. This is clear. I'll update this part.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+            logging.error(
+                f"`{file_path}' doesn't exist. Please make sure that the '{file_path_name}' argument is correct."
+            )
+            return None
```
same as above, replace None with False

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+        with open(json_path, "w") as fw:
+            json.dump(case_suite_in_template_format, fw, indent=4)
+
+    def read_case(self, file_name: str, case_suite: Dict) -> Dict:
```
Is this an API method or internal helper? The return type hint does not match the actual return

leijerry888 replied:
do not pass case_suite in this method, and only return newly added cases ids.

yunjoonjung-PNNL replied:
This is an internal helper method. The return type and arg/return value were fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+        ):
+            for case in loaded_cases["cases"]:
+                # check if there is any duplicated case. If so, don't add the case to `self.case_suite`
+                if all([loaded_cases != ele[1] for ele in self.case_suite.items()]):
```
seems like this is comparing many cases != one case. The intention is probably one vs one. maybe add a unit test to check this.

yunjoonjung-PNNL replied:
great catch! I fixed this. I'll also add this part in the unit test.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+                                "There is no JSON files in the given file_path."
+                            )
+                    else:
+                        logging.error(
```
this is also when a non-json file path is given

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,546 @@
+            file_path = ["wrong_file_path_type.json"]
+            vc = VerificationCase(case=None, file_path=file_path)
+            self.assertEqual(
+                "ERROR:root:The `file_path` argument's type must be <class 'str'>, but <class "
+                "'list'> is provided.",
```
one line?

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+                    # check if the directory exists
+                    if os.path.exists(file_path):
+                        # find all the JSON files in file_path directory
+                        json_file_names = glob.glob(os.path.join(file_path, "*.json"))
```
this is not tested in unit tests

yunjoonjung-PNNL replied:
Addressed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+                        # check if JSON file(s) exists.
+                        if len(json_file_names) != 0:
+                            for file in json_file_names:
+                                self.case_suite = self.read_case(
```
this will only load cases from the last file

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+                        if len(json_file_names) != 0:
+                            for file in json_file_names:
+                                self.case_suite = self.read_case(
+                                    file_path, self.case_suite
```
file_path should be file?

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+            # check 'file_path' type
+            if self.check_type("file_path", file_path, str):
+                # check if json or directory path is provided
+                if self.is_json_path_type(file_path):
+                    # check if 'file_path' exists
+                    if self.check_file("file_path", file_path):
+                        self.case_suite = self.read_case(file_path, self.case_suite)[0]
+                    else:
+                        return None
```
this is load_verification_cases_from_json

yunjoonjung-PNNL replied:
The difference is there is an else case in```if self.is_json_path_type(file_path):``` condition which doesn't belong to the ```load_verification_cases_from_json``` method. Thus, I believe we can't use the method in this case.

leijerry888 replied:
why load_verification_cases_from_json does not check if the path is json?

yunjoonjung-PNNL replied:
My apologies. I understood this correctly now and got this fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+
+    def load_verification_cases_from_json(
+        self, json_case_path: str = None
+    ) -> List[str]:
```
None is also possible

yunjoonjung-PNNL replied:
Good point! Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+        if self.check_type("json_case_path", json_case_path, str):
+            # check if `json_case_path` exists
+            if self.check_file("json_case_path", json_case_path):
+                self.case_suite, newly_added_hash = self.read_case(
```
this will eliminate previously loaded cases. Maybe add unit tests to after code fix to 1) make sure this won't happen and 2) also test that duplicate cases got correctly dropped.

yunjoonjung-PNNL replied:
Added to the unit test. I tested this part and any of the ```self.case_suite``` isn't eliminated. Please let me know if you want me to investigate further.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+
+    def save_case_suite_to_json(
+        self, json_path: str = None, case_ids: List = []
+    ) -> NoReturn:
```
there are returns

yunjoonjung-PNNL replied:
Which is None?

Question) In this case, is it correct to set the return type as ```UNION[NoReturn, None]```? Because there is a case there is no return?

leijerry888 replied:
No Return means this method does not return. E.g. it will raise exception within. return None is not NoReturn.

https://stackoverflow.com/a/3459131

yunjoonjung-PNNL replied:
I understand the difference between the ```None``` return and ```NoReturn```.

The gist of my question was considering the code structure, there could be a case where there is no return and a case where there is a ```None``` return because if the method is executed correctly, there is no return and if for example a wrong```json_path``` argument is provided, there is a None return.

So, is it correct to set the return type hint as ```UNION[NoReturn, None]```? Also, could you check the stackoverflow link? it seems like the link is not relevant for this discussion.

leijerry888 replied:
I did not see a case with NoReturn. Unless I missed something, when the method is executed correctly, it returns None although it was not written explicitly.

yunjoonjung-PNNL replied:
great point! Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+            loaded_cases = json.load(f)
+
+        newly_added_hash = []
+        newly_added_hash_append = newly_added_hash.append
```
this feels unnecessary

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+
+        # save case(s) based on `case_ids` arg
+        cases = []
+        cases_append = cases.append
```
this feels unnecessary

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+            return None
+
+        # organize the cases in the correct format
+        case_suite_in_template_format = {"cases": []}
+        case_suite_in_template_format_append = case_suite_in_template_format[
+            "cases"
+        ].append
+        for case in cases:
+            case_suite_in_template_format_append(case)
```
feels unnecessary, just:
```suggestion
        case_suite_in_template_format = {"cases": cases}
```

yunjoonjung-PNNL replied:
great point! Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+        base_case: Dict = None,
+        update_key_value: Dict = None,
+        keep_base_case: bool = False,
+    ) -> List[Dict]:
```
also can be None

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+
+        # deep-copy the base_case
+        generated_base_cases_list = []
+        generated_cases_list_append = generated_base_cases_list.append
```
similar to other occasions, no need to do this

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+        # update key/value based on `update_key_value` to `base_case`
+        def _update_key_helper(
+            update_key_value: Dict, casees: Dict, prev_level_key: str = None
+        ) -> Union[Dict, str]:
```
the actual return is a tuple

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+
+                elif isinstance(value, list):
+                    for idx, case in enumerate(casees):
+                        if key in [
+                            "no",
+                            "run_simulation",
+                            "expected_result",
+                            "datapoints_source",
+                            "verification_class",
+                        ]:
+                            case[key] = value[idx]
+                        elif key in ["idf", "idd", "weather", "output", "ep_path"]:
+                            case["simulation_IO"][key] = value[idx]
+
+                        elif key in ["subject", "variable", "frequency"]:
+                            case["datapoints_source"]["idf_output_variables"][
+                                prev_level_key
+                            ][key] = value[idx]
+
+                        elif prev_level_key == "parameters":
+                            case["datapoints_source"]["parameters"][key] = value[idx]
```
this is less ideal. Any change of schema would require change of this. Maybe we can use recursion and the updated key chain to help navigate the update. E.g. https://stackoverflow.com/a/14692747

leijerry888 replied:
also, we will need to handle the scenario of update keys not in base case

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+        self.save_verification_cases_to_json(json_path, cases)
+
+    @staticmethod
+    def create_verificaton_case_suite_from_base_case(
```
typo, should be create_verificaton_case_suite_from_base_case. please do a refactor as this typo is in all references and in md

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+
+        # update key/value based on `update_key_value` to `base_case`
+        def _update_key_helper(
+            update_key_value: Dict, casees: Dict, prev_level_key: str = None
```
cases: List ?

yunjoonjung-PNNL replied:
good catch! Fixed.

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+                "ep_path": str,
+            },
+            "expected_result": str,
+            "datapoints_source": {"idf_output_variables": dict, "parameters": dict},
```
this seems to require all cases have both idf output variables and parameters. In many verifications we don't use parameters

yunjoonjung-PNNL replied:
That's true, but when even if the ```parameters``` key isn't used in the ```new_library_verification_cases.json```, there is ```parameters``` key. Do we plan to make the ```parameters``` key optional in the future?

leijerry888 replied:
I would prefer to have them optional when they don't have any contents unless you feel it is very difficult to implement

yunjoonjung-PNNL replied:
Fixed. Now, the code works regardless of whether ```parameters``` exists or now.

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,546 @@
+            == False
+        )
+
+        # test when `run_simulation` key is wrong
```
comment seems to be off.

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,546 @@
+            )
+
+        # test when `run_simulation` key is wrong
+        case_wrong_run_simulation = {
```
one nit picky suggestion would be to have the wrong cases cases here and below to be modified from the correct cases so that no need to define all theses dicts from scratch, thus make code cleaner and issues more prominent

yunjoonjung-PNNL replied:
This is brilliant. Addressed,

leijerry888 commented on src/api/verification_case.py:
```diff
@@ -0,0 +1,430 @@
+    @staticmethod
+    def save_verification_cases_to_json(
+        json_path: str = None, cases: list = None
+    ) -> NoReturn:
```
save as above, return is None here

yunjoonjung-PNNL replied:
Fixed.

leijerry888 commented on tests/api/test_verification_case.py:
```diff
@@ -0,0 +1,546 @@
+        # when wrong json_path type is provided
+        with self.assertLogs() as logobs:
+            wrong_json_path = ["./wrong_json_path.json"]
+            VerificationCase(case=None, file_path=None).save_verification_cases_to_json(
```
to make sure this method is a static method, test without instantiate an object. This also applies to tests of other static methods.

yunjoonjung-PNNL replied:
Is there any way to use a staticmethod without object instantiation? Could you show me one example?

leijerry888 replied:
yes, just ClassName.method_name(args)

yunjoonjung-PNNL replied:
Good point! Fixed.

leijerry888 reviewed (changes requested):
Awesome work @yunjoonjung-PNNL . With your new tests, I did another round of review. Please see my comments and let me know if you have any questions

yunjoonjung-PNNL committed:
address PR comments(2)

yunjoonjung-PNNL committed:
Merge branch 'api_dev' into verification_case_api

yunjoonjung-PNNL committed:
formatting (black)

yunjoonjung-PNNL committed:
Merge branch 'verification_case_api' of https://github.com/pnnl/ANIMATE into verification_case_api

yunjoonjung-PNNL committed:
formatting with the latest black version

yunjoonjung-PNNL committed:
minor fix in error msg

yunjoonjung-PNNL committed:
fix read_case logic

yunjoonjung-PNNL committed:
fix constructor logic, get rid of `vc =` when not used

yunjoonjung-PNNL committed:
add a new testing data JSON file

yunjoonjung-PNNL commented:
@leijerry888 

I appreciate your comments and effort in this PR!

I addressed all the comments except:
i) using reduce function to traverse dict -> I'll revisit this this weekend as it takes a bit more time.
ii) I couldn't figure out how to turn on the type hint warning in Pycharm. Nonetheless, I believe I updated the typehint parts correctly, but I'll double-check this when I figure out how to turn on the type hint part in Pycharm (also this weekend).

Other than these, I believe this PR is ready to go through another round!

leijerry888 committed:
update verification case api constructor

leijerry888 committed:
change duplicate case check

leijerry888 committed:
remove redundant load cases test

leijerry888 committed:
update saving routine

leijerry888 committed:
refactor case structure validation

leijerry888 committed:
deep copy bug fix

leijerry888 committed:
new black 23

leijerry888 reviewed (approved):
@yunjoonjung-PNNL I added a few commits to reflect my review comments. please see if they make sense. I am merging this back so Jeremy and I can work on other API with this. If you have questions / suggestions. We can change it with a new PR.

md needs to be updated too.
